src/test/rgw/test_d4n_filter.cc

```diff
@@ -2759,6 +2952,437 @@ TEST_F(D4NFilterFixture, CopyReplaceVersionedObjectWrite)
+
+    // Make sure delete marker is written to backend
+    objEnabled->set_instance(deleteMarker);
+    auto next = dynamic_cast<rgw::sal::FilterObject*>(objEnabled.get())->get_next();
+    EXPECT_EQ(next->load_obj_state(env->dpp, optional_yield{yield}), 0);
+    EXPECT_EQ(next->exists(), 1);
+
+    conn->cancel();
+    testBucket->remove(env->dpp, true, optional_yield{yield});
+    DriverDestructor driver_destructor(static_cast<rgw::sal::D4NFilterDriver*>(driver));
+  }, rethrow);
+
+  io.reset();
+  io.run();
+}
+
+TEST_F(D4NFilterFixture, VersionedDeleteAfterCleaning)
+{
+  env->cct->_conf->d4n_writecache_enabled = true;
+  env->cct->_conf->rgw_d4n_cache_cleaning_interval = 0.5;
+  const std::string testName = "VersionedDeleteAfterCleaning";
+  const std::string bucketName = "/tmp/d4n_filter_tests/dbstore-default_ns.1";
+  std::vector<std::string> instances;
+ 
+  net::spawn(io, [this, &testName, &bucketName, &instances] (net::yield_context yield) {
+    init_driver(yield);
+    create_bucket(testName, yield);
+    testBucket->get_info().bucket.bucket_id = bucketName;
+    
+    for (int i = 0; i <= 1; ++i) {
+      std::string instance;
+      put_version_enabled_object(testName, instance, yield); // Upload two versions 
+      instances.push_back(instance);
+    }
+
+    dynamic_cast<rgw::d4n::LFUDAPolicy*>(d4nFilter->get_policy_driver()->get_cache_policy())->save_y(null_yield);
+  }, rethrow);
+
+  io.run_for(std::chrono::seconds(2)); // Allow cleaning cycle to complete
+
+  net::spawn(io, [this, &testName, &bucketName, &instances] (net::yield_context yield) {
+    dynamic_cast<rgw::d4n::LFUDAPolicy*>(d4nFilter->get_policy_driver()->get_cache_policy())->save_y(optional_yield{yield});
+
+    std::error_code err;
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + url_encode(bucketName, true) + "/" + TEST_OBJ + testName + "/" + instances[1], err), true);  
+    std::string oid = instances[1] + "#0#" + std::to_string(ofs);
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + url_encode(bucketName, true) + "/" + TEST_OBJ + testName + "/" + oid, err), true);  
+
+    std::unique_ptr<rgw::sal::Object::DeleteOp> del_op_enabled = objEnabled->get_delete_op();
+    objEnabled->set_instance(instances[1]); // Latest version
+    EXPECT_EQ(del_op_enabled->delete_obj(env->dpp, optional_yield{yield}, rgw::sal::FLAG_LOG_OP), 0);
+    
+    // Make sure deleted instance is deleted from backend
+    {
+      auto next = dynamic_cast<rgw::sal::FilterObject*>(objEnabled.get())->get_next();
+      EXPECT_EQ(next->load_obj_state(env->dpp, optional_yield{yield}), -2);
+    }
+
+    objEnabled->set_instance(instances[0]);
+
+    // First instance should still available in backend
+    {
+      auto next = dynamic_cast<rgw::sal::FilterObject*>(objEnabled.get())->get_next();
+      EXPECT_EQ(next->load_obj_state(env->dpp, optional_yield{yield}), 0);
+      EXPECT_EQ(next->exists(), 1);
+    }
```

Comment:
similar comment - try to verify what is the latest version id, maybe use next->iterate?